### PR TITLE
Documentation change for lazyproperty to reflect setup of setter/deleter

### LINKED ISSUE #4471: lazyproperty: Documentation misleading
I tried the ``lazyproperty`` from the utils.decorators today and it is works as expected but one thing is very strange. The documentation states that

If a setter for this property is defined, it will still be possible to manually update the value of the property, if that capability is desired.

Well but at least with python 3.5 the setter/deleter are always created. To stay with the example in the docstring of the decorator
```
class LazyTest(object):
    @lazyproperty
    def complicated_property(self):
        print('Computing the value for complicated_property...')
        return 42
lt = LazyTest()
lt.complicated_property
#Computing the value for complicated_property...
#42
lt.complicated_property
#42
# New
lt.complicated_property = 10
lt.complicated_property
#10
del lt.complicated_property
lt.complicated_property
#Computing the value for complicated_property...
#42
```
I didn't expected this behaviour from the docstring and was searching for some time now where my code went wrong. If this is intentional I would suggest altering the docstring or if it wasn't intentional maybe there is a way to alter it. I haven't checked but maybe this may also be something that was different in an older python version. Or am I misunderstanding the documentation?

### PR DESCRIPTION
**Pull request by MSeifert04**
Fixes #4471

### CODE REVIEW

**Comment by MSeifert04** [comment]
Actually this documentation change is not really accurate since the setter/deleter are not overridden if a setter and deleter are implemented since in the code there is no if/else for custom ``fset``, ``fdel`` in https://github.com/astropy/astropy/blob/master/astropy/utils/decorators.py#L528 and 533. But writing ``extended`` instead of ``overridden`` is also not true since the ``setter`` is overridden if it doesn't do anything wrong. Would it be more accurate to write the ``setter`` can be ``overridden`` and the ``deleter`` ``extended``?

**Comment by mhvk** [comment]
@MSeifert04 - having looked at the code again, I see that the default actions are indeed not automatically removed when using one's own setter and deleter. I think it may be best to rewrite the text to something like:

By default, a setter and deleter are used which simply overwrite and delete, respectively, the value stored in ``__dict__``. Any user-specified setter or deleter is executed before executing these default actions. The one exception is that the default setter is not run if the user setter already sets a new value in ``__dict__`` and returns that value.

File: astropy/utils/decorators.py [hunk 1]
```diff
@@ -493,8 +493,8 @@ class lazyproperty(property):
     the ``print`` statement is not executed.  Only the return value from the
     first access off ``complicated_property`` is returned.
 
-    If a setter for this property is defined, it will still be possible to
-    manually update the value of the property, if that capability is desired.
+    The lazyproperty also sets up a simple setter and deleter for the property
+    which can be overridden by own setter and deleter methods.
```
Comment (embray) [inline astropy/utils/decorators.py]: I don't think this is much less confusing.  See my comment on the original issue.

**Comment by embray** [comment]
I think instead this should update the documentation to describe, in prose, exactly what the existing code does.

**Comment by mhvk** [comment]
@embray - what do you think about my alternative suggestion above?

**Comment by embray** [comment]
@mhvk I didn't see that before.  That sounds right.  There just also needs to be some words about how the getter stores the computed value in `__dict__` in the first place.

**Comment by MSeifert04** [comment]
@mhvk: Your text is definitly better and I tried around a bit to find suitable ways of making it customizable:

Disabling the setter:
```
from astropy.utils.decorators import lazyproperty
class A(object):
    @lazyproperty
    def testprop(self):
        return 10

    @testprop.setter
    def testprop(self, value):
        raise AttributeError('cannot set attribute')
```
Implementing setter logic
```
class A(object):
    @lazyproperty
    def testprop(self):
        return 10

    @testprop.setter
    def testprop(self, value):
        if not isinstance(value, int):
            # Everything that doesn't raise an error will be saved
            raise TypeError('cannot set attribute with non-integers')
```
If one wants to alter the saved value it's quite hard (because of https://github.com/MSeifert04/astropy/blob/fix-4471/astropy/utils/decorators.py#L523)
```
class A(object):
    @lazyproperty
    def testprop(self):
        return 10

    @testprop.setter
    def testprop(self, value):
        value = np.array(value) # change value in some arbitary way
        self.__dict__['testprop'] = value # necessary
        return value # also necessary
```
Now, I only have to find a way to make that into a text. :-) (which I can't, I've tried several times now)

**Comment by mhvk** [comment]
I think one change we could make that would simplify the description is that anything else than `None` returned by a user setter is simply stored in `__dict__` -- I think this makes sense since it leaves how the value is stored as an implementation detail. It also think it is safe in not breaking any existing code.

**Comment by embray** [comment]
@mhvk I think I tried that once but I actually did need to be able to save `None` as well.  I do like the idea that whatever is returned by the setter is automatically saved in `__dict__` though.  So there would need to be some other way for a custom setter to explicitly signal to *not* store in `__dict__`.  Either some kind of special sentinel value (other than `None`--perhaps some sentinel object attached to the `lazyproperty` class itself), or by raising a particular exception.

**Comment by embray** [comment]
Maybe for now though this issue should just focus on documenting the existing behavior and not changing it.

I've actually been meaning to bringing this up with python-ideas, because I think `lazyproperty`, or something like it, is useful enough to be worth adding to the standard library.  Maybe the python-ideas folks, if they like the idea at all, can help with the bike-shedding on that...

**Comment by mhvk** [comment]
@embray - setting to `None` using a user-defined setter and inputting a value other than `None` is not possible as is, since it is checked that the output from `fset` is not `None`:
```
    def __set__(self, obj, val):
        obj_dict = obj.__dict__
        if self.fset:
            ret = self.fset(obj, val)
            if ret is not None and obj_dict.get(self._key) is ret:
                # By returning the value set the setter signals that it took
                # over setting the value in obj.__dict__; this mechanism allows
                # it to override the input value
                return
        obj_dict[self._key] = val
```
Of course, this does not prevent `myclass.lazyproperty = None` to work, but that would be OK in my alternative as well:
```
    def __set__(self, obj, val):
        if self.fset:
            ret = self.fset(obj, val)
            if ret is not None:
                # override the value to be set.
                val = ret
        obj.__dict__[self._key] = val
```
I like how this simplifies the description: any user-defined setter has to return the value to be actually set, by default, it will just be whatever is passed in.

**Comment by mhvk** [comment]
Saw your latest message only after I wrote. if I were to start over, I'd probably just unconditionally store the output of `self.fset`; but as is, I think my suggestion makes it sufficiently more logical that one might as well go for it. That said, this particular PR is about documentation, and even just using some variation of the text I suggested above would be an improvement, so I'll leave this up to @MSeifert04.

**Comment by MSeifert04** [comment]
Well I have read about some implementations of lazyproperties:
https://github.com/mitsuhiko/werkzeug/blob/master/werkzeug/utils.py#L35
https://github.com/jackmaney/lazy-property/blob/master/lazy_property/__init__.py#L4
https://pypi.python.org/pypi/lazy
https://gist.github.com/wonderbeyond/df700e559f7114822d15#file-lazy_property-py-L8

But that's maybe a bit more than I actually wanted: My only problem with the documentation was that it implements a setter and deleter while the documentation says it wouldn't.

**Comment by MSeifert04** [comment]
@mhvk This time I was writing while you commented. I think you are right, having your documentation would be better than mine. Shall I change it or do you want to create an own PR?

**Comment by embray** [comment]
Ah, right.  I agree it might be simpler if by default it always set the return value of `fset` into `__dict__`.

I think part of the issue is that in `astropy.io.fits` there are some cases where there are setters for `self.data` that need to manually update `__dict__` *first*, because within the setter there are calls to other methods which reference `self.data`.  So it's necessary to make sure that `__dict__['data']` has already been updated--it can't rely on the default functionality happening first.

A better implementation would be to allow the setter to be a *generator*.  It could `yield` the value that should be stored in `__dict__['data']`, but still allow additional code to be run by the setter afterwards.  I think I'll make a PR for that.

**Comment by embray** [comment]
By the way, if I implemented the above then most of the discussion about the internal details could be moot (though it's maybe still worth having anyways).

For the deleter it could also accept a generator (though in this case any yielded value would be ignored--it just allows it to have pre- and post-deletion code to run).

**Comment by embray** [comment]
Hmm, it seems that one other use case from `astropy.io.fits` is, in the `setter`, checking if a value for that attribute has already been set (for example, `if 'data' in self.__dict__: ...`).  The idea here is that a custom setter can still be employed to do additional checks and/or transformations on the value before storing it, while still checking if a previous value for that attribute has been cached.  I don't see a better way to do this than manually checking in `__dict__`, and maybe that's fine.  It just needs to be documented better as we've already discussed.

**Comment by MSeifert04** [comment]
@mhvk If you don't mind I just pushed your suggestion with a slight addition.

@embray If you want to alter the way the ``lazyproperty`` works this will probably take a bit longer since it's being used. Should I include some more documentation about how to use the custom ``setter`` (like in https://github.com/astropy/astropy/pull/4476#issuecomment-171042360) or will you change the implementation of ``lazyproperty`` soon?

**Comment by mhvk** [comment]
@MSeifert04 - I think the change looks good and believe we should just merge is now, returning to this if/when we change `lazyproperty`.

**Comment by embray** [comment]
Yeah, might as well go with this for now.  I spent some time yesterday playing with a new implementation but I forget where I left that--I was getting pretty tired.  We'll see.

I changed the milestone to v1.0.9 since this is just a documentation improvement that applies to all actively developed versions of Astropy.

**Comment by mhvk** [comment]
OK, merging this! Not sure if rewriting `lazyproperty` is worth raising a new issue for...

**Pull request merged**